cachelib/cachebench/workload/BinaryKVReplayGenerator.h

```diff
@@ -0,0 +1,289 @@
+  std::atomic<uint32_t> incrementalIdx_{0};
+
+  // A sticky index assigned to each stressor threads that calls into
+  // the generator.
+  folly::ThreadLocalPtr<uint32_t> stressorIdx_;
+
+  // Vector size is equal to the # of stressor threads;
+  // stressorIdx_ is used to index.
+  std::vector<std::unique_ptr<StressorCtx>> stressorCtxs_;
+
+  // Pointer to request object used to carry binary
+  // request data
+  std::vector<Request*> requestPtr_;
+
+  // Class that holds a vector of pointers to the
+  // binary data
+  BinaryFileStream binaryStream_;
+
+  std::thread genWorker_;
+
+  // Used to signal end of file as EndOfTrace exception
+  std::atomic<bool> eof{false};
+
+  // Stats
+  std::atomic<uint64_t> parseError = 0;
+  std::atomic<uint64_t> parseSuccess = 0;
+
+  void genRequests(folly::Latch& latch);
+
+  void setEOF() { eof.store(true, std::memory_order_relaxed); }
+  bool isEOF() { return eof.load(std::memory_order_relaxed); }
+
+  inline StressorCtx& getStressorCtx(size_t shardId) {
+    XCHECK_LT(shardId, numShards_);
+    return *stressorCtxs_[shardId];
+  }
+
+  inline StressorCtx& getStressorCtx() {
+    if (!stressorIdx_.get()) {
+      stressorIdx_.reset(new uint32_t(incrementalIdx_++));
+    }
+
+    return getStressorCtx(*stressorIdx_);
+  }
+
+  inline Request* getRequestPtr(size_t shardId) {
+    XCHECK_LT(shardId, numShards_);
+    return requestPtr_[shardId];
+  }
+
+  inline Request* getRequestPtr() {
+    if (!stressorIdx_.get()) {
+      stressorIdx_.reset(new uint32_t(incrementalIdx_++));
+    }
+
+    return getRequestPtr(*stressorIdx_);
+  }
+};
+
+inline void BinaryKVReplayGenerator::genRequests(folly::Latch& latch) {
+  bool init = true;
+  uint64_t nreqs = 0;
+  binaryStream_.setOffset(fastForwardCount_);
+  auto begin = util::getCurrentTimeSec();
+  while (!shouldShutdown()) {
+    try {
+      BinaryRequest* req = binaryStream_.getNextPtr();
+      auto key = req->getKey(binaryStream_.getKeyOffset());
+      XDCHECK_LT(req->op_, 11);
+      nreqs++;
+      auto shardId = getShard(key);
+      auto& stressorCtx = getStressorCtx(shardId);
+      auto& reqQ = *stressorCtx.reqQueue_;
+      while (!reqQ.write(req) && !stressorCtx.isFinished() &&
+             !shouldShutdown()) {
+        // ProducerConsumerQueue does not support blocking, so use sleep
+        if (init) {
+          latch.count_down();
+          init = false;
+        }
+        std::this_thread::sleep_for(
+            std::chrono::microseconds{checkIntervalUs_});
+      }
+      if (nreqs >= preLoadReqs_ && init) {
+        auto end = util::getCurrentTimeSec();
+        double reqsPerSec = nreqs / (double)(end - begin);
+        XLOGF(INFO, "Parse rate: {:.2f} reqs/sec", reqsPerSec);
+        latch.count_down();
+        init = false;
+      }
+    } catch (const EndOfTrace& e) {
+      if (init) {
+        latch.count_down();
+      }
+      break;
+    }
+  }
+
+  setEOF();
+}
+
+const Request& BinaryKVReplayGenerator::getReq(uint8_t,
```

Comment:
this function (and the entire file) looks pretty similar to KVReplayGenerator.h Can't we unify them? Perhaps make this (and other) function a template (where either BinaryRequest* or std::unique_ptr<ReqWrapper> is the arg)?

Reply:
It's actually pretty hard because we don't have the same underlying request queue logic in the binary trace version.